tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/LoopEncoder.scala

```diff
@@ -0,0 +1,267 @@
+package at.forsyte.apalache.tla.pp.temporal
+
+import at.forsyte.apalache.tla.lir._
+
+import javax.inject.Singleton
+import com.typesafe.scalalogging.LazyLogging
+import at.forsyte.apalache.tla.typecomp._
+import at.forsyte.apalache.tla.pp.temporal.ScopedBuilderExtensions._
+import at.forsyte.apalache.tla.pp.temporal.DeclUtils._
+import at.forsyte.apalache.tla.pp.temporal.utils.builder
+import at.forsyte.apalache.tla.lir.transformations.TransformationTracker
+
+/**
+ * A class for adding loop logic to a TLA Module: An extra variable InLoop is added, which, in each step, can
+ * nondeterministically be set to true. When the variable is set to true, the current values of variables are stored in
+ * copies of the variables (e.g. the value of foo is stored in an auxiliary variable loop_foo). When InLoop is true, the
+ * system is inside of the loop.
+ *
+ * It is ensured that the execution loops, i.e. the state at the start of the loop is the same as the state at the end
+ * of the execution, by considering only traces where the variable values at the last state match the ones that were
+ * stored at the start of the loop.
+ *
+ * @author
+ *   Philip Offtermatt
+ */
+@Singleton
+class LoopEncoder(tracker: TransformationTracker) extends LazyLogging {
+  import LoopEncoder.NAME_PREFIX
+
+  val boolTag = Typed(BoolT1)
+
+  val inLoopDecl = TlaVarDecl(LoopEncoder.IN_LOOP_NAME)(boolTag)
+  val inLoop = builder.declAsNameEx(inLoopDecl)
+  val inLoopPrime = builder.primeVar(inLoopDecl)
+
+  /** For each variable foo, creates a declaration of an auxiliary variable loop_foo */
+  def createLoopVariables(originalVariables: Seq[TlaVarDecl]): Seq[TlaVarDecl] = {
+    originalVariables.map(varDecl => createLoopVariableForVariable(varDecl))
+  }
+
+  def createLoopVariableForVariable(varDecl: TlaVarDecl): TlaVarDecl = {
+    TlaVarDecl(s"${NAME_PREFIX}${varDecl.name}")(varDecl.typeTag)
+  }
+
+  /**
+   * Takes a variable foo and its corresponding loop variable loop_foo, and transforms init into
+   *
+   * newInit == init /\ loop_foo = foo
+   */
+  def addLoopVarToInit(varDecl: TlaVarDecl, loopVarDecl: TlaVarDecl, init: TlaOperDecl): TlaOperDecl = {
+    conjunctExToOperDecl(
+        builder.eql(builder.declAsNameEx(loopVarDecl), builder.declAsNameEx(varDecl)),
+        init,
+        tracker,
+    )
+  }
+
+  /**
+   * Transforms init into
+   *
+   * newInit ==
+   *
+   * /\ init
+   *
+   * /\ inLoop = FALSE
```

Comment:
InLoop = true in state i means state i-1 is the first state on the loop.
It's the way it's done in the paper (see p22 of https://lmcs.episciences.org/2236)

Note that the behaviour like this neatly disallows single-states loops unless they are allowed by having stuttering.

If conversely InLoop being true at state i means state i is the first state of the loop, you could immediately end the execution, since the copy vars are equal to the vars by definition.

Reply:
Will be fixed in https://github.com/informalsystems/apalache/pull/1823